### PR TITLE
Updates and documents testing tools

### PR DESCRIPTION
**Pull request by urimandujano**
Improves the state of testing tools in ActionProviderTools. The `mocks.py` module is greatly slimmed down and provides a new `TokenChecker` mock which can be used to replace `TokenChecker` instances.  The addition of `fixtures.py` introduces fixtures that mock Globus Auth integration for the two Flask toolk, APTBlueprint and Flask Helpers.

Examples are updated to demonstrate using the fixtures to test apps without a Globus Auth Client ID, secret, or token.

Docs are updated with to reflect the new fixtures and updated mocks as well as with information on when/how to use each fixture and when to use the mocks instead.

Adds a deprecation warning to the testing patches, mostly because the `flask_blueprint_tokenchecker_patch` is not as good as the `apt_blueprint_noauth` fixture (the fixture mocks only the blueprint-under-test's TokenChecker whereas the patch mocks the TokenChecker globally for all blueprint's.). Also the `flask_api_helpers_tokenchecker_patch` is effectively the same as the `flask_helpers_noauth` fixture, except in both cases the fixtures are more convenient and easier to work with.

Adds tests validating the new mocks. "Dogfoods" the new fixtures by using them in the toolkit test's for the Flask helpers.

### CODE REVIEW

File: docs/source/toolkit/testing.rst [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+.. code-block:: python
+
+    def test_introspection_endpoint(client):
+        response = client.get("/")
+        assert response.status_code == 200
+
+.. note:: 
+    
+    The :code:`flask_helpers_noauth` fixture will patch the TokenChecker in a
```
Comment (jimPruyne) [inline docs/source/toolkit/testing.rst:118]: Seems less likely for others than how we do things, but an excellent note to make it clear.

File: docs/source/toolkit/testing.rst [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+Mock AuthState
+--------------
+
+An *AuthState* represents a requestor's authentication status and Globus Auth
+information. Every request should have its token validated via the
+*TokenChecker*'s :code:`check_token` method, which in turns generates an
+*AuthState* object.
+
+During testing, it is desirable to not need to provide valid tokens with every
```
Comment (jimPruyne) [inline docs/source/toolkit/testing.rst:151]: "no not need to" perhaps awkward? (Am I sounding like an English teacher...)

File: docs/source/toolkit/testing.rst [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+request. Use the :code:`mock_authstate` mock to generate a stubbed out
+*AuthState* object that won't validate requestor properties against Globus Auth.
+This is most useful when used in a patch as the return value for the
+*TokenChecker*'s :code:`check_token` method:
+
+.. code-block:: python
+
+    from unittest import mock
```
Comment (jimPruyne) [inline docs/source/toolkit/testing.rst:159]: If we're using pytest, does using its monkeypatch module make at least as much sense as using `unittest.mock`? Do we have concerns about mixing unittest and pytest?

File: docs/source/toolkit/testing.rst [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+    from unittest import mock     
+
+    import pytest
+    from globus_action_provider_tools.testing.mocks import mock_authstate
+
+    ...
+
+    @pytest.fixture
+    def client():
```
Comment (jimPruyne) [inline docs/source/toolkit/testing.rst:167]: With monkeypatch, I think this would look roughly like:
```python
@pytest.fixture
def client(monkeypatch):
    monkeypatch.settattr("globus_action_provider_tools.authentication.TokenChecker.check_token",
            mock_authstate)
```

So pretty similar.

File: docs/source/toolkit/testing.rst [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            "globus_action_provider_tools.authentication.TokenChecker.check_token",
+            return_value=mock_authstate(),
+        ):
+            yield app.test_client()
+
+The example above creates a fixture which can be used to create a client that
+can make unauthenticated HTTP requests against an Action Provider. 
+
+
+Mock TokenChecker
```
Comment (jimPruyne) [inline docs/source/toolkit/testing.rst:178]: This seems like a really good approach for non-flask apps? Should we push this description forward so people get to it before the other approach?

File: docs/source/toolkit/testing.rst [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            app.config["TESTING"] = True
+            yield app.test_client()
+
+
+.. note::
+
+    This example will only work if there's a function or method that is used to
+    create the TokenChecker instance. It demonstrates how you can patch a
+    function or a method to return the Mock TokenChecker. Internally, the Mock
+    TokenChecker will generate the :ref:`mock-authstate` objects described
```
Comment (jimPruyne) [inline docs/source/toolkit/testing.rst:212]: Ah, this kinda makes sense to have the other description first. Perhaps makes my previous comment not so good.

File: examples/apt_blueprint/tests/test_action_provider.py [hunk 7]
```diff
@@ -3,21 +3,19 @@
...
-    with flask_blueprint_tokenchecker_patch:
-        app = create_app()
-        yield app.test_client()
+def client(apt_blueprint_noauth):
+    # Patch the blueprint BEFORE attaching it to an app so that the Provider can
+    # start without valid credentials AND requests can be made without supplying
+    # a valid token
+    apt_blueprint_noauth(aptb)
+    app = create_app()
+    return app.test_client()
```
Comment (jimPruyne) [inline examples/apt_blueprint/tests/test_action_provider.py:18]: The examples in the docs use `yield` here its `return`. GIven that there's no code after the `yield` above, there's probably no actual difference (?). Maybe nice to be consistent. I think we get used to just using `yield` in fixtures sometimes.

File: examples/watchasay/tests/test_action_provider.py [hunk 8]
```diff
@@ -21,20 +21,17 @@
...
-        app = create_app()
-        app.config["TESTING"] = True
-        yield app.test_client()
+def client(flask_helpers_noauth):
+    # Patch the app using the flask_helpers_noauth fixture so that the Provider
+    # can start up without valid credentials AND requests can be made without
+    # supplying a valid token
+    app = create_app()
+    app.config["TESTING"] = True
+    yield app.test_client()
```
Comment (jimPruyne) [inline examples/watchasay/tests/test_action_provider.py:34]: See above for yield vs return

File: examples/whattimeisitrightnow/tests/test_action_provider.py [hunk 9]
```diff
@@ -31,14 +31,20 @@
...
 
 @pytest.fixture
 def client():
-    app.config["TESTING"] = True
-    yield app.test_client()
+    with mock.patch(
+        "globus_action_provider_tools.authentication.TokenChecker.check_token",
+        return_value=mock_authstate(),
+    ):
+        from examples.whattimeisitrightnow.app.app import app
```
Comment (jimPruyne) [inline examples/whattimeisitrightnow/tests/test_action_provider.py:44]: is this import here related to how the mock is getting installed? Its kinda a funky notation...

File: globus_action_provider_tools/testing/fixtures.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    Auth integration. The fixture returns a function to which the instance
+    should be supplied as a parameter:
+
+    i.e. apt_blueprint_noauth(aptb)
+    """
+
+    def _apt_blueprint_noauth(aptb: ActionProviderBlueprint):
+        # Manually remove the function that creates the internal token_checker
+        for f in aptb.deferred_functions:
+            if f.__name__ == "_create_token_checker":
```
Comment (jimPruyne) [inline globus_action_provider_tools/testing/fixtures.py:22]: Given how this is mucking with the internals, really good to provide the function for people as we wouldn't want them to have to figure this out.

File: globus_action_provider_tools/testing/fixtures.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+
+@pytest.fixture(scope="module")
+def flask_helpers_noauth():
+    """
+    A fixture designed to mock the Globus Auth integration in an Flask app
+    created using the api_helpers. Simply using this fixture will allow creating
+    the mocked app.
+    """
+    with patch(
```
Comment (jimPruyne) [inline globus_action_provider_tools/testing/fixtures.py:38]: So, I see that my suggestion around monkeypatch might have repercussions down in the internal implementation. Maybe take a quick look to see if it does it in a nice way that keeps us consistent in the pytest world since we're already assuming it. If it makes it any bit harder, don't worry about it.

File: globus_action_provider_tools/testing/fixtures.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        "globus_action_provider_tools.flask.api_helpers.TokenChecker.check_token",
+        return_value=mock_authstate(),
+    ):
+        yield
```
Comment (jimPruyne) [inline globus_action_provider_tools/testing/fixtures.py:42]: one nice thing about monkeypatch is you don't have to do this yield. Since the monkeypatch itself is a fixture, it handles that.

File: globus_action_provider_tools/testing/mocks.py [hunk 13]
```diff
@@ -1,45 +1,41 @@
-from typing import NamedTuple
-from unittest.mock import PropertyMock, patch
+from unittest.mock import Mock
```
Comment (jimPruyne) [inline globus_action_provider_tools/testing/mocks.py:1]: So, here, unittest.mock makes sense to me as it doesn't really on any outside packages to be installed so people not going flask-y can also not go pytest-y.

Either I'm over thinking this, or under thinking it. I'm not quite sure...

File: globus_action_provider_tools/testing/mocks.py [hunk 14]
```diff
@@ -1,45 +1,41 @@
...
+    auth_state.check_authorization.return_value = True
+    auth_state.introspect_token.return_value = None
 
-    auth_client = ConfidentialAppAuthClient(None, None)
-    auth_state = AuthState(auth_client, None, None)
-    auth_state.check_authorization = mock_check_authorization
-    auth_state.primary_identity = mock_primary_identity
-    auth_state.introspect_token = mock_introspect_token
-    auth_state.bearer_token = "MOCK_BEARER_TOKEN"
+    return auth_state
```
Comment (jimPruyne) [inline globus_action_provider_tools/testing/mocks.py:31]: This looks much cleaner, nice!

File: globus_action_provider_tools/testing/patches.py [hunk 15]
```diff
@@ -1,7 +1,19 @@
+import warnings
 from unittest.mock import patch
 
 from globus_action_provider_tools.testing.mocks import mock_authstate
 
+warnings.warn(
```
Comment (jimPruyne) [inline globus_action_provider_tools/testing/patches.py:6]: nice

File: tests/test_flask_helpers/conftest.py [hunk 16]
```diff
@@ -28,7 +32,7 @@
 
 
 @pytest.fixture()
-def aptb_app(auth_state):
+def aptb_app(apt_blueprint_noauth, auth_state):
```
Comment (jimPruyne) [inline tests/test_flask_helpers/conftest.py:35]: fixture here is so nice to read compared to the `with patch`

**Review by jimPruyne — COMMENTED** [review]
Looks really good. Really nice to see both the functionality and particularly the full docs!

A bit of general wording stuff in docs, but pretty minor.

Some quesitons about using unittest.mock vs pytest.monkeypatch in places? But, I'm open to whichever solution between those two seems best to you.

File: docs/source/toolkit/testing.rst [hunk 17]
```diff
@@ -0,0 +1,213 @@
+Testing
+=======
+
+An Action Provider is closely tightle with Globus Auth (see
```
[the remark this answers is not in the document]
Comment (urimandujano) [reply docs/source/toolkit/testing.rst:4]: Wow, I can't remember what I was going for. Switching to "integrated" since it sounds like a fancier word.

File: docs/source/toolkit/testing.rst (hunk 1 above)
Comment (urimandujano) [reply docs/source/toolkit/testing.rst:118]: Making note of things that have bitten me before so that other's don't get hurt (:

File: docs/source/toolkit/testing.rst (hunk 3 above)
Comment (urimandujano) [reply docs/source/toolkit/testing.rst:159]: I think the only concern is providing a uniform non-confusing story for folks who are ramping up with testing/using the tools. `unittest` and `pytest` play pretty well with each other so I'm not too worried about them interfering with each other (unless there are cases and I just haven't run across them yet). 

There are some gotchas with using pytest's `monkeypatch` fixture because it's function-scoped. The fixtures provided here are using `patch` and `mock` directly so they can be used in fixtures regardless of their scope. I like the idea of using `monkeypatch` in the examples since it's syntax is clearer than patch's.

File: docs/source/toolkit/testing.rst (hunk 4 above)
Comment (urimandujano) [reply docs/source/toolkit/testing.rst:167]: Thanks for the working example!

File: docs/source/toolkit/testing.rst (hunk 5 above)
Comment (urimandujano) [reply docs/source/toolkit/testing.rst:178]: I haven't had the chance to test this on a non-Flask app to see how we could just mock the `TokenChecker` import so I'm not sure how/if it'll work. I think mocking the import would work but I'm not quite ready to claim that it's possible or that it's a good idea.

For now I've added it as a piece that can be used and outlined a particular way of using it that I know works (via the provided fixtures) we start experimenting with other frameworks or want to revisit how framework agnostic our tools are, this would be something to work on pushing forward.

File: examples/apt_blueprint/tests/test_action_provider.py (hunk 7 above)
Comment (urimandujano) [reply examples/apt_blueprint/tests/test_action_provider.py:18]: I've also seen yield most often used in test fixtures so `yield` it is!

File: examples/whattimeisitrightnow/tests/test_action_provider.py (hunk 9 above)
Comment (urimandujano) [reply examples/whattimeisitrightnow/tests/test_action_provider.py:44]: Yea it has to do with how the `TokenChecker` gets instantiated at import time... The example could probably be refactored to have a slightly cleaner way of getting mocked but I was hoping not to spend too much time refactoring examples.

File: globus_action_provider_tools/testing/fixtures.py (hunk 11 above)
Comment (urimandujano) [reply globus_action_provider_tools/testing/fixtures.py:38]: Yea I think using `monkeypatch` internally would impose some limitations on how our libraries fixtures could be used (i.e. they could only be using in a certain scope). Looking at this again I've actually decided to broaden the scope of our fixtures to "session" (which is as I understand it the broadest scope possible) so that they can be used in fixtures of any scope. 

Using patches at this level maintains allows some extra flexibility for use cases, which may or may not be needed at this point.

File: globus_action_provider_tools/testing/mocks.py (hunk 13 above)
Comment (urimandujano) [reply globus_action_provider_tools/testing/mocks.py:1]: I'm not sure how to prove that it's over or under thinking (: My hope in providing base mocks was that they'd be able to be composed by higher level tools depending on what needed to be accomplished. I think it worked nicely that the `fixtures` end up dogfooding the `mocks` and that the `tokenchecker` mock dogfoods the `auth_state` mock. From some of my udpates to the examples, they also seem to fit in well when using `monkeypatch`.

**Comment by urimandujano** [comment]
Thank you for the review! It was good for me to step back and think through the strengths of `unittest's` tools and `monkeypatch`. I think I've convinced myself that at the lowest level, unittest tools provide the most flexibility (which is welcome since we're trying to provide general purpose testing tools) but if that flexibility isn't necessary, `monkeypatch` is the way to go (useful for the testing examples + realworld usage).

Really interesting to me is that it seems like there are lots of people trying to make sense of the two mock methods and opinions vary widely: https://github.com/pytest-dev/pytest/issues/4576.

**Review by jimPruyne — APPROVED** [review]
After discussion, this all looks good.
